### ISSUE TITLE
[macOS] Cannot run in fullscreen

### ISSUE THREAD

**Issue by shinra-electric** [dhewm/dhewm3#436]
When changing resolution through the system menu, changing resolution will work after exiting and restarting the app.

However, if I set fullscreen to yes and restart, it will return to a default small window. 
![Screenshot 2022-01-06 at 22 53 53](https://user-images.githubusercontent.com/50119606/148464252-422e30ee-523c-4415-9f6e-7977f6cd46c5.png)
![Screenshot 2022-01-06 at 22 54 13](https://user-images.githubusercontent.com/50119606/148464262-1755d726-d412-49b9-a619-b74a927e66f9.png)

In addition, if I set a resolution for windowed mode, it doesn't seem right. For example, the screenshot below is supposed to be 1440p, but it's clearly not taking up the whole screen...

![Screenshot 2022-01-06 at 22 59 54](https://user-images.githubusercontent.com/50119606/148464608-6335730b-8567-435b-abf9-1284533a1786.png)

**Comment by DanielGibson** [comment]
For sake of completeness, some comments you made in the MacSourcePorts Discord:
> Ah, it seems it might be an SDL bug, but combined with the fact I'm using two monitors.

> ----- Warnings -----
> during dhewm 3 initialization...
> WARNING: Current display mode isn't requested display mode
> WARNING: Likely SDL bug #4700, trying to work around it..
> WARNING: Still in wrong display mode: 2560x1440 instead of 1920x1200

> the 1440p one is the one I'm running it on. My other monitor is 1920x1200

and your logfile: [Dhewm3_Terminal_Log.txt](https://github.com/dhewm/dhewm3/files/7825313/Dhewm3_Terminal_Log.txt)

An additional log for fullscreen would be nice, but I can imagine that it's related to the same problem (dhewm3 and/or SDL is confused about which display is which)

**Comment by DanielGibson** [comment]
> Hakase: First run: set to 1920x1200 windowed mode and restart
> Hakase: second run: toggle fullscreen and restart
> Hakase: third run: Still opened in small window, so quit
> Hakase: One warning is WARNING: Still in wrong display mode: 1920x1200 instead of 1680x1050
> Hakase: I don't know where it got the 1680x1050, that was not selected anywhere (by me) 
> Hakase: Question: Could the settings in the options be setting the window to the wrong size? I selected 1920x1200, but when it restarted in windowed mode it was not close to the full screen size. It may have changed it to 1680x1050 instead?

The log: [Dhewm3_log.txt](https://github.com/dhewm/dhewm3/files/7847300/Dhewm3_log.txt)

**Comment by DanielGibson** [comment]
@shinra-electric I just pushed a commit that will log even more information, can you rebuild from latest git and try again?

**Comment by shinra-electric** [comment]
Update: 

1st run - Open (small window). Set resolution to 2560x1440, clicked apply and quit
2nd run - Window opened at roughly 1600x1200 (4:3 ratio, black bars at the sides). Selected full-screen to yes and quit.
3rd run - it opened in the small window

[dhewm3log-1stRun.txt](https://github.com/dhewm/dhewm3/files/7889183/dhewm3log-1stRun.txt)
[dhewm3log-2ndRun.txt](https://github.com/dhewm/dhewm3/files/7889184/dhewm3log-2ndRun.txt)
[dhewm3log-3rdRun.txt](https://github.com/dhewm/dhewm3/files/7889185/dhewm3log-3rdRun.txt)

**Comment by DanielGibson** [comment]
I just can't reproduce this.
If I set the resoultion to 2560x1440 in the menu, `r_mode` gets set to `23`, like it should (when I open the console and just enter `r_mode`, it says it's 23). After restarting the game, I get a window in that resolution and r_mode is still set to 23.
Could you repeat steps 1 and 2, and in step 1, after clicking "apply", enter `r_mode` in the console? And then do the same thing again after restarting the game.
(you don't have to enter `listmodes` again though)

**Comment by DanielGibson** [comment]
does this problem still exist?

**Comment by shinra-electric** [comment]
Yes, I just made a new build and it exhibits the same behaviour.

**Comment by DanielGibson** [comment]
Does it also happen with the MacSourcePorts build? IIRC that one has something set in some startup script or Info.plist or whatever for HighDPI-support, which might help.

If that doesn't help, please try the following:
1. Start dhewm3, set resolution to 2560x1440, click apply DO NOT QUIT YET
2. Open console (Shift+Esc), enter `r_mode`
3. Quit game, save the terminal log (to attach it here later)
4. Start dhewm3 again
5. Open console, enter `r_mode` again
6. Quit the game, again save the terminal log

Then please post both terminal logs here.

**Comment by shinra-electric** [comment]
Just to clarify, for Dhewm3 I normally use the MSP build. Earlier today I built it myself just in case you recently made any changes to the code that might have addressed the issue. 

The following logs are from the MSP build: 
[Terminal Saved Output 1.txt](https://github.com/dhewm/dhewm3/files/8703023/Terminal.Saved.Output.1.txt)
[Terminal Saved Output 2.txt](https://github.com/dhewm/dhewm3/files/8703024/Terminal.Saved.Output.2.txt)

**Comment by DanielGibson** [comment]
Thanks!
So you set the mode to 2560x1440 in the menu, right?
And then afterwards (before any restart, in the first log) r_mode is 16 (which is 1920x1200) - that's weird.
What does the menu show after restart? Still 2560x1440 or 1920x1200?

What happens if you set `r_mode 23` in the console and restart the game afterwards?

**Comment by shinra-electric** [comment]
Just to add, after I set to `r_mode 23`, I noticed that the menu says `r_custom`
![Screenshot 2022-05-16 at 21 38 46](https://user-images.githubusercontent.com/50119606/168678789-9c66c1e1-cce3-439b-b26f-618a0fe471ff.png)

**Comment by DanielGibson** [comment]
Somehow the internal sorting of the modes (as displayed in the menu) must be messed up, but I have no idea why that would happen - the code looks correct, it works as expected on my (Windows and Linux) machines and there is absolutely nothing Mac-specific about that code..

Could be a bug in the Apple-Clang compiler or in Apple's libc (maybe in `qsort()`)

Can you find out if this also happens with x86_64 binaries (either natively on an Intel Mac or in Rosetta)?

**Comment by shinra-electric** [comment]
Yes, I have an Intel Mac I can test on, though I probably won't do it tonight.

**Comment by DanielGibson** [comment]
Thanks for testing that!

Can you check out and build the `debug-res-list` branch: https://github.com/dhewm/dhewm3/tree/debug-res-list
And then test if the bug still happens and post the console output of one run?

**Comment by shinra-electric** [comment]
Here you go. Tested on Monterey 12.4 with an M1.

1st run: loaded with whatever settings I had previously, screen size was 640x480. 
Used the menu to change to 1440p and quit. 

2nd run: Opened with a medium-sized window (not 1440p). 
Used the menu to turn on fullscreen mode and quit. 

3rd run: Opened in 480p. Quit. 

[Terminal Saved Output.txt](https://github.com/dhewm/dhewm3/files/8747434/Terminal.Saved.Output.txt)

**Comment by hbokh** [comment]
Same issue here on a MBP M2. Binary built yesterday:
```
dhewm3 1.5.4pre.1305 macosx-arm64 Jun 28 2024 16:14:02 using SDL v2.30.4
SDL video driver: cocoa
```

The issue happens here when starting the game from Terminal / iTerm2 and then it checks "where the cursor is at".
When the terminal is on the MacBook screen:
```
SDL detected 2 displays:
 0: 2560x1440 at (0, 0) to (2560, 1440)
 1: 1512x982 at (2560, 409) to (4072, 1391)
Will use display 1 because mouse cursor is at (2881, 1192).
```
starting a windowed game in `r_mode 3` even though `autoexec.cfg` has it set to 23.

When the terminal is on the external display:
```
SDL detected 2 displays: 
 0: 2560x1440 at (0, 0) to (2560, 1440)
 1: 1512x982 at (2560, 409) to (4072, 1391)
Will use display 0 because mouse cursor is at (1404, 922).
```
it starts the game in the mode as it is setup in `autoexec.cfg`.

Just as with Quake III Arena (`quake3e`) and Quake 2 (`yquake2`) I would like the game to always open on the external display (`display 0`), never minding where the cursor is at.

**Comment by ralph303** [comment]
Anyone got fullscreen working on MacOS?
I use the latest MacSourcePorts build (Version 1.5.5) together with the data from the GOG game.
But fullscreen isn't working, no matter what I try.
I use MacOS 15.7.7 and a MacBook Air M4.

**Comment by DanielGibson** [comment]
I can't debug this because I don't own a Mac.

I'd suggest asking the MacSourcePorts folks for help (maybe in their Discord if you use that), there you're also more likely to find someone who has tried getting fullscreen working (successfully or not).

To be clear, I don't mind this problem being discussed here and I hope that eventually a fixed that I can merge can be found - I just can't help much myself.

**Comment by DanielGibson** [comment]
OK one thing worth trying is: Open the dhewm3Settings menu (either with F10 or if that doesn't work open the console with Shift+Esc and enter `dhewm3Settings`).
Then go to `Video Options`, select `[x] Fullscreen Desktop Mode` and click `Apply`.

That should make the game run in a fullscreen window at the current display resolution

**Comment by ralph303** [comment]
Thanks for your quick replies, I think I got it working.
You first have to lookup the exact resolution of your Macs Display (for example in the System Report or typing "system_profiler SPDisplaysDataType" in the Terminal).
In my case the MacBook M4 has a resolution of 2560 x 1664.
Then just select the proper resolution in the games system menu (2560x1660 will be fine) and check the fullscreen option.
If the correct resolution is not selectable, then choose "custom", exit the game and set the values manually in the config file.

If the game won't start in widescreen automatically, then this setting might be useful:
seta r_aspectRatio "2" // Set aspect ratio: 0=4:3 (default), 1=16:9 (widescreen), 2=16:10 (ultrawide)

It works great now on the MacBook M4, thanks a lot for this port and for supplying a MacOS version of it.

**Comment by DanielGibson** [comment]
r_aspectratio really shouldn't have anything to do with it and should remain at the default of `-1`. It does not affect **at all** if a window can be created, only (to some degree) if the game will look right or somehow skewed.

Fullscreen Desktop should just work, and the menu will also show additional helpful information like that virtual resolution and the real resolution (in case of HighDPI/Retina)

**Comment by ralph303** [comment]
Yes, Fullscreen Desktop works even with lower resolutions!
If you want to use the fullscreen option in the games menu you have to set the exact resolution of your display, otherwise it won't work and reset the resolution to 640x480.

**Comment by DanielGibson** [comment]
Fullscreen Desktop ignores the configured resolution and just uses the current screen resolution, by turning dhewm3 into a borderless fullscreen window. It is a dhewm3 feature, which is why the old Doom3 menu doesn't support it.

IMHO using the dhewm3 Settings Menu is generally a good idea for all settings that can be configured there, because it regards dhewm3-specific features.
